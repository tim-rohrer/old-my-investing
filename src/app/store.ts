import { configureStore, ThunkAction, Action } from "@reduxjs/toolkit";
import counterReducer from "../features/counter/counterSlice";
import portfolioReducer from "../features/portfolio/portfolioSlice";

export const reducer = {
  counter: counterReducer,
  portfolio: portfolioReducer,
};

export const store = configureStore({
  reducer,
});

export type RootState = ReturnType<typeof store.getState>;

export type AppDispatch = typeof store.dispatch;

export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
