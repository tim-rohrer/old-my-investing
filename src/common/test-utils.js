// test-utils.js
import React from "react";
// import { render as rtlRender } from '@testing-library/react';
import { render } from "@testing-library/react";
import { configureStore } from "@reduxjs/toolkit";
import { createStore } from "redux";
import { Provider } from "react-redux";
import {
  initialStateOriginal as reducerInitialState,
  reducer,
  initialStateOriginal,
} from "../app/store";

// const store = configureStore({ reducer });

export function renderWithRedux(
  ui,
  {
    initialState = initialStateOriginal,
    store = configureStore({ reducer }),
  } = {}
) {
  return {
    ...render(<Provider store={store}>{ui}</Provider>),
    // adding `store` to the returned utilities to allow us
    // to reference it in our tests (just try to avoid using
    // this to test implementation details).
    store,
  };
}
// export function customRender(
//   ui,
//   {
//     initialState: RootState = reducerInitialState,
//     rtlStore,
//     ...renderOptions
//   } = {}
// ) {
//   function Wrapper({ children }) {
//     const rtlStore = configureStore({ reducer });
//     return <Provider store={rtlStore}>{children}</Provider>;
//   }
//   return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
// }

// re-export everything
export * from "@testing-library/react";

// override render method
// export { customRender as render };
