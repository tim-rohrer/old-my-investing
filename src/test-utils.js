// test-utils.js
import React from 'react'
import { render as rtlRender } from '@testing-library/react'
import { configureStore } from '@reduxjs/toolkit'
// import { createStore } from 'redux'
import { Provider } from 'react-redux'
import { reducer } from './app/store'
// import { initialState as reducerInitialState, reducer } from './reducer'
const reducerInitialState = {}

function render(
  ui,
  {
    initialState: RootState = reducerInitialState,
    rtlStore,
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    const rtlStore = configureStore({ reducer })
    return <Provider store={rtlStore}>{children}</Provider>
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions })
}

// re-export everything
export * from '@testing-library/react'

// override render method
export { render }