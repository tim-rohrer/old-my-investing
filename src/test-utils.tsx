// test-utils.tsx
import * as React from "react"
import { Provider } from "react-redux"
import { createStore } from "redux"

import { queries, Queries } from "@testing-library/dom"
import { render as rtlRender, RenderOptions, RenderResult } from "@testing-library/react"

import { rootReducer, State, Store } from "./app/store"

type ExtraOptions = {
  initialState?: State;
  store?: Store;
};

const render = <
  Q extends Queries = typeof queries,
  C extends Element | DocumentFragment = HTMLElement
>(
  ui: React.ReactElement,
  options: ExtraOptions & RenderOptions<Q, C> = {} //:
): RenderResult<Q, C> & { store: Store } => {
  const {
    initialState = undefined, // reducerInitialState,
    store = createStore(rootReducer, initialState),
    ...renderOptions
  } = options;

  type Props = {
    children?: React.ReactNode
  }

  const Wrapper = ({ children }: Props) => {
    return <Provider store={store}>{children}</Provider>;
  };

  return {
    ...rtlRender(ui, { wrapper: Wrapper, ...renderOptions }),
    store,
  };
};

// re-export everything
export * from "@testing-library/react";

// override render method
export { render };
