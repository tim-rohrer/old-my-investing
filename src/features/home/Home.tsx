import React, { FC } from "react";
import {
  AppBar,
  IconButton,
  InputBase,
  Toolbar,
  Typography,
} from "@material-ui/core";
import { useStyles } from "../../common/useStyles";
import MenuIcon from "@material-ui/icons/Menu";
import SearchIcon from "@material-ui/icons/Search";
import { Autocomplete } from "@material-ui/lab";
import { useDispatch, useSelector } from "react-redux";
import { AppDispatch } from "../../app/store";
import { appIsThinking, selectAppIsLoaded } from "../system/systemSlice";
import {
  fetchFMPCompaniesSymbolsList,
  FMPCompanySymbol,
  selectCompaniesSymbols,
} from "../securities/securitiesSlice";
import {
  fetchFMPCompanyProfileBySymbol,
  FMPRequestObject,
} from "../companyAnalysis/companyProfilesSlice";
import { CompanyProfile } from "../companyAnalysis/CompanyProfile";
import useCompanyProfile from "../companyAnalysis/useCompanyProfile";

type HomeProps = {
  title: string;
};

/**
 * @description The home or main container for My Investing prototype.
 * This level is where top level data coordination occurs.
 */
const Home: FC<HomeProps> = ({ title }: HomeProps) => {
  const classes = useStyles();
  const dispatch: AppDispatch = useDispatch();

  const [companySymbol, setCompanySymbol] = React.useState<string>("DG");
  const isAppLoaded = useSelector(selectAppIsLoaded);
  const companiesSymbolsList = useSelector(selectCompaniesSymbols);
  const currentCompanyProfile = useCompanyProfile(companySymbol);

  React.useEffect(() => {
    if (!isAppLoaded) {
      dispatch(appIsThinking(true));
      dispatch(
        fetchFMPCompaniesSymbolsList({
          requestType: "companiesSymbolsList",
        })
      );
    }
  }, [isAppLoaded, dispatch]);

  // React.useEffect(() => {
  //   if (companySymbol) {
  //     useCompanyProfile(companySymbol);
  //   } else console.log("Beats me!");
  // }, [companySymbol]);

  return (
    <div className={classes.root}>
      <AppBar position="sticky">
        <Toolbar>
          <IconButton
            edge="start"
            className={classes.menuButton}
            color="inherit"
            aria-label="open drawer"
          >
            <MenuIcon />
          </IconButton>
          <Typography className={classes.title} variant="h6" noWrap>
            {title}
          </Typography>
          <Autocomplete
            id="symbol-search"
            autoHighlight
            options={companiesSymbolsList}
            getOptionLabel={(option) => `${option.name} (${option.symbol})`}
            style={{ width: 400 }}
            value={companySymbol}
            onChange={(event: any, newValue: string) => {
              setCompanySymbol(newValue);
            }}
            noOptionsText="No securities found"
            renderInput={(params) => {
              const { InputLabelProps, InputProps, ...rest } = params;
              return (
                <div className={classes.search}>
                  <div className={classes.searchIcon}>
                    <SearchIcon />
                  </div>
                  <InputBase
                    {...params.InputProps}
                    {...rest}
                    placeholder="Symbol or Name"
                    classes={{
                      root: classes.inputRoot,
                      input: classes.inputInput,
                    }}
                  />
                </div>
              );
            }}
          />
        </Toolbar>
      </AppBar>
      {currentCompanyProfile !== undefined ? (
        <CompanyProfile {...currentCompanyProfile} />
      ) : (
        "Loading..."
      )}
    </div>
  );
};

export default Home;
