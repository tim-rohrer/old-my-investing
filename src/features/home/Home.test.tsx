import React from "react";
import { render, cleanup, screen, fireEvent } from "../../test-utils";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom/extend-expect";
import { configureStore } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import { Store, rootReducer as reducer } from "../../app/store";
import Home from "./Home";

let store: Store;
let component: any;

beforeEach(() => {
  store = configureStore({
    reducer,
  });
  component = render(
    <Provider store={store}>
      <Home title="test app" />
    </Provider>
  );
});

afterEach(cleanup);

describe("home", () => {
  it("renders a snapshot", () => {
    expect(component.container).toMatchSnapshot();
  });

  it("provides a search box for entering a symbol or company name", () => {
    expect(screen.getByPlaceholderText(/Symbol/)).toBeInTheDocument();
  });

  it("renders the Company Profile after the user enters a symbol", () => {
    // screen.debug();
    (global as any).document.createRange = () => ({
      setStart: () => {},
      setEnd: () => {},
      commonAncestorContainer: {
        nodeName: "BODY",
        ownerDocument: document,
      },
    });
    // console.log(screen.queryByText("Company Profile"));
    // expect(screen.queryByText("Company Profile")).toBeNull();
    const input = screen.getByPlaceholderText("Symbol or Name");
    userEvent.type(input, "SPY{enter}");
    // expect(screen.getByText("Company Profile")).toBeInTheDocument();
  });
});
