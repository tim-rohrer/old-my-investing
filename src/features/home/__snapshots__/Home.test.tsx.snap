// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`home renders a snapshot 1`] = `
<div>
  <div
    class="MuiBox-root css-i9gxme"
  >
    <header
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionSticky css-7e0s21-MuiPaper-root-MuiAppBar-root"
    >
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1t29gy6-MuiToolbar-root"
      >
        <button
          aria-label="open drawer"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart MuiIconButton-sizeLarge css-184mzhy-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="MenuIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <div
          class="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap css-1dvstpp-MuiTypography-root"
        >
          test app
        </div>
        <div
          class="css-96uuyl"
        >
          <div
            class="css-1iz4iln"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="SearchIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
              />
            </svg>
          </div>
          <div
            aria-expanded="false"
            class="MuiAutocomplete-root css-1bu2955-MuiAutocomplete-root"
            role="combobox"
            style="width: 400px;"
          >
            <div
              class="MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-19ipv7b-MuiInputBase-root"
            >
              <input
                aria-autocomplete="list"
                autocapitalize="none"
                autocomplete="off"
                class="MuiInputBase-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-yz9k0d-MuiInputBase-input"
                id="symbol-search"
                placeholder="Symbol or Name"
                spellcheck="false"
                type="text"
                value=""
              />
              <div
                class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
              />
            </div>
          </div>
        </div>
      </div>
    </header>
  </div>
</div>
`;
